### PR TITLE
fix: Improve frequency of Remote MCDU updates

### PR DESCRIPTION
**Pull request by frankkopp**
### Problem Description
The current implementation of the Remote MCDU sends updates with the full content of the MCDU page VERY often. 
There is a regular update every 500ms independent of any changes to the page.
Also the every internal MCDU update is forwarded to the Remote MCDU and some pages even update every frame even though this is not necessary on this page (no changes or changes which would work with a slower update frequency). 

### This PR
Improving the update handling for the MCDU to reduce the number of updates sent to the Remote MCDU.

- Removed the 500ms update - tests show that this is not required at all 
- Changed page updates to the minimum required update frequency by adapting update categories (Fast, Medium, Dyn, Default, Slow) and changing the page updates to a common update method. 
- Added a loading message to the "Closest Airports" page as initial load can take several seconds. 
- Added "NOT YET IMPLEMENTED" to AIRPORT key and on DIR TO page for ABEAM, RADIAL in and out

Additional (fixes/changes after review and first teste):
- fixes the console.log error appearing when not connected to the MCDU Server
- checking power regularly as turning off and on does not send any event for updates
- refactored the regular tasks (instead setInterval() this uses the onUpdate() method and an updateThrottler)
- prevented input from events and keyboard when MCDU has no power
- fixes AOC RCVD MSGS and AOC SENT MSGS pages to update correctly when new messages are available
- fixes pagination for these two pages
- sending free text or departure clearance and selecting another page before sending was done made the MCDU jump back to the respective pages. Added checks to prevent that. 

Discord username (if different from GitHub): Cdr_Maverick#6475

### Priority 1:
- MCDU works as before 
- Remote MCDU works as before
- Check AOC FREE TEXT AND DEPART REQ to work as before
- Check RECEIVED MESSAGES and SENT MESSAGES page to update correctly and to paginate correctly with more than 5 messages

Pay attention to page updates on the MCDU and Remote MCDU - actions and pages which should change the content need to still work. 

E.g. W&B pax loading, INIT page, GPS Monitor (has seconds), MESSAGES pages update, etc.

Doing a full flight allows for testing flight phase changes and maybe a turn around test would complete the picture. 

### Priority 2:
- Updates to the Remote MCDU should happen rarely - most page update every 1-2s - only a few faster
- Loading the Data - Airport page for the first time should show a "Loading..." message. 
- Pressing AIRPORT key should trigger a "NOT YET IMPLEMENTED" message in the scratchpad
- On DIR TO page pressing ABEAM, RADIAL in, or RADIAL out LSK should trigger a "NOT YET IMPLEMENTED" message in the scratchpad
- Remote MCDU should behave correctly when power is shut off (display off, content in sync when power on again)
- When power off MCDU in cockpit and Remote MCDU do not accept input any longer (they did before)

#### Firefox:
Open the MCDU in a Browser on a PC (needs PC browser). 
Press F12 to get the browser dev tools up.
Go to "Network" and reload the page in the browser.
You should now see a list of loaded files. 
Click on bundle.js
image
Click on "Response" to see updates to the page from the websocket. 
image

#### Chrome + Edge:
image

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by tracernz** [comment]
If I can ask you a favour, assuming you haven't already done it... Can you catch the connection error and stop it spamming the dev console?
image

**Comment by frankkopp** [comment]
> If I can ask you a favour, assuming you haven't already done it... Can you catch the connection error and stop it spamming the dev console? image

Will look into it - I haven't got this?

**Comment by tracernz** [comment]
> Will look into it - I haven't got this?

You will only get it if you don't have the server running.

**Comment by frankkopp** [comment]
> 



> If I can ask you a favour, assuming you haven't already done it... Can you catch the connection error and stop it spamming the dev console? image

Checked it, did some research and tried a few things. 
1. This error message has been there before (current master) but only every 5sec. I have changed it to 1sec as I wasn't aware of this issue. I will change it back to 5sec as it was before. 
2. There seems to be no easy way to avoid this error message. It seems to be deliberately designed this way. 

Maybe you know a way?

So we will have to live with the error message every 5sec if no MCDU Server is running.

**Review by tracernz — APPROVED** [review]

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6626 
Tier of Testing  : 1
Date             : 15/02/2022

Testing Process: 
MCDU works as before - **OK**
Remote MCDU works as before - **OK**
Pay attention to page updates on the MCDU and Remote MCDU - actions and pages which should change the content need to still work. E.g. W&B pax loading, INIT page, GPS Monitor (has seconds), etc. - **OK**

Updates to the Remote MCDU should happen rarely - most page update every 1-2s - only a few faster - **OK**
Loading the Data - Airport page for the first time should show a "Loading..." message. - **OK**
Pressing AIRPORT key should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - **OK**
On DIR TO page pressing ABEAM, RADIAL in, or RADIAL out LSK should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - **OK**

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

Media:       
image
Most pages update every one or two seconds.
image
Not yet implemented message

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js
Comment (tracernz) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:282]:
```suggestion
            this.updateRequest = false;
            if (this.pageRedrawCallback) {
```

Comment (frankkopp) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:282]: ups :)

**Comment by frankkopp** [comment]
> If I can ask you a favor, assuming you haven't already done it... Can you catch the connection error and stop it spamming the dev console? image

btw. this is solved now :) thanks to @BlueberryKing

**Review by BlueberryKing — APPROVED** [review]

**Comment by ShadowFrost3** [comment]
Quality Assurance Tester Report


Object of testing: #6626 
Tier of Testing:1
Date: 02/20/2022

Testing Process: 
Priority 1:
MCDU works as before - checked
Remote MCDU works as before - checked
Pay attention to page updates on the MCDU and Remote MCDU - actions and pages which should change the content need to still work. E.g. W&B pax loading, INIT page, GPS Monitor (has seconds), etc. - checked 

Priority 2:
Updates to the Remote MCDU should happen rarely - most page update every 1-2s - only a few faster - checked
Loading the Data - Airport page for the first time should show a "Loading..." message. - checked 
Pressing AIRPORT key should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - checked 
On DIR TO page pressing ABEAM, RADIAL in, or RADIAL out LSK should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - checked 

Notes- 

Should the remote MCDU turn off when the aircraft is off?
Remote MCDU doesn't always sync to aircraft MCDU when aircraft is powered on. 

Negatives: 
Testing Results: Passed assuming the above are not regressions

**Comment by frankkopp** [comment]
@ShadowFrost3 I have pushed an update that should fix the power on/off de-sync.
This needs some re-testing and maybe there is a more elegant way to solve this.
Will look into it a bit but maybe you can test if this already solves your issue.

**Comment by frankkopp** [comment]
As Hoppie has been merged since this PR has been sent to review I found a few issues with the PDC, Rcvd, Sent and Freetext pages. Will fix them first and then ping for another round of testing. 
Until then I have marked this as "Do Not Merge"

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js
Comment (tracernz) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js:9]: `page = Math.max(0, Math.min(Math.floor((messages.length - 1) / 5), page));`

Comment (frankkopp) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js:9]: @tracernz 
Made this change - testing it during the day.
Did you look at my other changes as well? I'm a bit concerned as this PR is a bit larger as I anticipated and needs good review and testing.

**Comment by svengcz** [comment]
Tested the Hoppie-releated pages on functionality:
- Tested Freetext: **OK**
- Tested PDC: **OK**
- Tested auto-update of "Messages received": **OK**
- Tested auto-update of "Messages sent": **OK**

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js
Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js]: Looks good to me. One more QA pass to merge (since it failed for ShadowFrost)?

Comment (frankkopp) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_MessagesSent.js]: Yes, maybe two :) it is an important component

**Comment by ShadowFrost3** [comment]
Quality Assurance Tester Report


Object of testing: #6626
Tier of Testing:1
Date: 28/02/2022

Testing Process: 

SIM UPDATE 8

Priority 1:
MCDU works as before  - checked
Remote MCDU works as before - checked
Pay attention to page updates on the MCDU and Remote MCDU - actions and pages which should change the content need to still work. E.g. W&B pax loading, INIT page, GPS Monitor (has seconds), etc. - checked

Priority 2:
Updates to the Remote MCDU should happen rarely - most page update every 1-2s - only a few faster - not rechecked 
Loading the Data - Airport page for the first time should show a "Loading..." message. - checked
Pressing AIRPORT key should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - checked
On DIR TO page pressing ABEAM, RADIAL in, or RADIAL out LSK should trigger a "NOT YET IMPLEMENTED" message in the scratchpad - checked

Syncs through restart - checked 
Syncs through aircraft power up/down - checked 

Negatives: 
Testing Results: Passed

Note SU8 related - there were many oddities related to FMS/FCU functions but unlikely to be a result of this PR

**Pull request merged**
